java/org/contikios/cooja/Simulation.java

```diff
@@ -150,68 +155,133 @@ public Simulation(Cooja cooja, long seed) {
-
-        /* Handle one simulation event, and update simulation time */
-        nextEvent = eventQueue.popFirst();
-        assert nextEvent != null : "Ran out of events in eventQueue";
-        assert nextEvent.time >= currentSimulationTime : "Event from the past";
-        currentSimulationTime = nextEvent.time;
-        nextEvent.event.execute(currentSimulationTime);
-
-        if (stopSimulation) {
-          isRunning = false;
-          var realTimeDuration = System.currentTimeMillis() - lastStartRealTime;
-          var simulationDuration = getSimulationTimeMillis() - lastStartSimulationTime;
-          logger.info("Runtime: {} ms. Simulated time: {} ms. Speedup: {}",
-                      realTimeDuration, simulationDuration,
-                      ((double) simulationDuration / Math.max(1, realTimeDuration)));
-        }
+      /* Remove all motes */
+      Mote[] motes = getMotes();
+      for (Mote m: motes) {
+        doRemoveMote(m);
       }
-    } catch (RuntimeException e) {
-    	if ("MSPSim requested simulation stop".equals(e.getMessage())) {
-    		/* XXX Should be*/
-    		logger.info("Simulation stopped due to MSPSim breakpoint");
-    	} else {
-
-    		logger.fatal("Simulation stopped due to error: " + e.getMessage(), e);
-    		if (!Cooja.isVisualized()) {
-    			/* Quit simulator if in test mode */
-    			System.exit(1);
-    		} else {
-    		  String title = "Simulation error";
-    		  if (nextEvent.event instanceof MoteTimeEvent) {
-    		    title += ": " + ((MoteTimeEvent)nextEvent.event).getMote();
-    		  }
-          Cooja.showErrorDialog(title, e, false);
-    		}
-    	}
+    }, "sim");
+    this.simulationThread.start();
+  }
+
+  /*
+   * Set if the simulation is running or not.
+   * May only be called by the simulation thread to notify about its state.
+   */
+  private void setRunning(boolean isRunning) {
+    this.isRunning = isRunning;
+    if (isRunning) {
+      /* Simulation starting */
+      speedLimitLastRealtime = lastStartRealTime = System.currentTimeMillis();
+      speedLimitLastSimtime = lastStartSimulationTime = getSimulationTimeMillis();
+    } else {
+      var realTimeDuration = System.currentTimeMillis() - lastStartRealTime;
+      var simulationDuration = getSimulationTimeMillis() - lastStartSimulationTime;
+      logger.info("Runtime: {} ms. Simulated time: {} ms. Speedup: {}",
+                  realTimeDuration, simulationDuration,
+                  ((double) simulationDuration / Math.max(1, realTimeDuration)));
     }
-    isRunning = false;
-    simulationThread = null;
-    stopSimulation = false;
 
-    cooja.updateProgress(true);
-    this.setChanged();
-    this.notifyObservers(this);
+    synchronized (lock) {
+      lock.notifyAll();
+    }
+
+    cooja.updateProgress(!isRunning);
+    setChanged();
+    notifyObservers(this);
+
+    if (!Cooja.isVisualized() && !isRunning) {
+      // Remove simulation when it stopped without GUI
+      cooja.doRemoveSimulation(false);
```

Comment:
Don't make this conditional on headless mode, if the code belongs here it should always be enabled.

Reply:
The purpose is to remove the simulation after the simulation has stopped but only when Cooja runs without GUI (runs in "test mode"). Cooja is waiting for the simulation to be removed before exiting or continuing to run the next simulation. But this is probably not the right place, and I will consider where this should be handled.

Reply:
Well, both `doQuit` and `loadSimulationConfig` removes the simulation.

(Maybe they shouldn't, and the simulation should be removed somewhere else, but that should be fixed in a separate PR after this has landed.)

Reply:
Well, without this Cooja will wait for ever for the simulation to be removed and never proceed to exit or run another simulation. But I agree it should be handled better. This is one of the reasons this PR is still a draft.